jax/_src/array.py

```diff
@@ -700,54 +700,79 @@ def make_array_from_callback(
+    arrays: Sequence of ``jax.Array``\s that are each single device addressable. ``len(arrays)``
+      must equal ``len(jax.local_devices())`` and the shape of each array must be the same. For multiprocess code,
+      the each process will call with a different ``arrays`` argument that corresponds to that processes' data.
 
   Returns:
-    A ``jax.Array`` from a sequence of ``jax.Array``\s on a single device.
+    A global ``jax.Array``, sharded as ``sharding``, with shape equal to ``shape``, and with per device
+      contents matching ``arrays``.
 
-  Example:
+  Examples:
+
+    In this single process example, we use ``make_array_from_single_device_arrays`` to create an
+    a global array.
 
     >>> import math
     >>> from jax.sharding import Mesh
     >>> from jax.sharding import PartitionSpec as P
     >>> import numpy as np
     ...
+    >>> mesh_rows = 2
+    >>> mesh_cols =  jax.device_count() // 2
+    ...
     >>> global_shape = (8, 8)
-    >>> global_mesh = Mesh(np.array(jax.devices()).reshape(2, 4), ('x', 'y'))
+    >>> global_mesh = Mesh(np.array(jax.devices()).reshape(mesh_rows, mesh_cols), ('x', 'y'))
     >>> sharding = jax.sharding.NamedSharding(global_mesh, P('x', 'y'))
     >>> inp_data = np.arange(math.prod(global_shape)).reshape(global_shape)
     ...
     >>> arrays = [
-    ...     jax.device_put(inp_data[index], d)
-    ...     for d, index in sharding.addressable_devices_indices_map(global_shape).items()]
+    ...    jax.device_put(inp_data[index], d)
+    ...        for d, index in sharding.addressable_devices_indices_map(global_shape).items()]
     ...
     >>> arr = jax.make_array_from_single_device_arrays(global_shape, sharding, arrays)
-    >>> arr.addressable_data(0).shape
-    (4, 2)
-
-    In multi-process case, if the input is process local and data parallel
-    i.e. each process receives a different part of the data, then you can use
-    `make_array_from_single_device_arrays` to create a global jax.Array
+    >>> assert arr.shape == (8,8)
+    # arr.shape is (8,8) regardless of jax.device_count()
+
+    When using multiple processes, the most common case is the input is process local and data parallel.
+    (Equivalently, we assume each process receives a different part of the data and the
+    total number of rows is at least one per device.) In this case, the following recipe will use
+    `make_array_from_single_device_arrays` to create a global jax.Array. This recipe is very typical
+    when creating data loading pipelines.
+
+    >>> sharding = jax.sharding.NamedSharding(global_mesh, P(('x', 'y'),))
+    >>> rows_per_device = 2
+    >>> feature_length = 32
+    >>> per_device_shape = (rows_per_device, feature_length)
+    >>> global_shape = (jax.device_count() * per_device_shape[0], ) + per_device_shape[1:]
+    >>> data_generator = lambda : np.arange(math.prod(per_device_shape)).reshape(per_device_shape)
```

Comment:
why do we need a generator?

I think we should keep the `split` code since that is what users will have to do right?

Reply:
The data_generator is nice because it is very suggestive of how to use the API when integrating. There is literally no reason we couldn't replace it with dummy data.

Are you sure they will have to split? My opinion is they can just replace data_generator with their per chip generation function and go...

Reply:
Reading your two questions together ... I feel like they're related!

My writing of this example is basically "how do you lift a per-chip data pipeline into a global data pipeline".

Reply:
Decision was made: yes generator but a whole host generator.